contributors/design-proposals/ipvs-proxy.md

```diff
@@ -0,0 +1,152 @@
+
+### Port remapping 
+
+There are 3 proxy mode in ipvs, NAT, IPIP and DR. Only NAT mode support port remapping.
+
+### NodePort type service support
+
+For NodePort type service, IPVS proxy will take all accessable IPs in a host as the virtual IP of ipvs service. Specifically, accessable IP excludes `lo`, `docker0`, `vethxxx`, `cni0`, `flannel0`, etc. Currently, we assume they are IPs bound to `eth{i}`.
+
+We people visit `NodeIP:NodePort`, it will hit the VIP of ipvs service. IPVS proxy doesn't need to assign Node IP to dummy interface.
+
+### LoadBalancer type service support
+
+For LoadBalancer type service, IPVS proxy will take the loadbalancer's ingress IP as the virtual IP of ipvs service. 
+
+Since there is a need of supporting access control for ingress IP. IPVS proxy will make use of iptables to do access control, Like
+
+```shell
+-I KUBE-SERVICES -d {ingress.IP} --dport {service.Port} -s {LoadBalancerSource} -j ACCEPT
+```
+
+When the packet reach the end of firewall chain and does not get DNATed, IPVS proxy will drop it, iptables rule like
+
+```shell
+-A KUBE-SERVICES -d {ingress.IP} --dport {service.Port} -j KubeMarkDropChain
+```
+
+### Support service cluster IP
+
+IPVS proxy will assign service cluster IP to the dummy interface on each cluster node. Specifically,  it will create an alias on dummy interface for each cluster IP. Delete a service will trigger the deletion of alias device.
+
+### Support service external IP
+
+IPVS proxy will not create alias device for external IPs of a service. It only create a ipvs service whose VIP corresponding to external IP.
+
+### Session affinity
+
+ipvs support client IP session affinity, it's called persistent connection. It's the `-p` flag in ipvsadm, like
+
+```shell
+ipvsadm -A -t 10.244.1.100:8080 -s rr -p [timeout]
+```
+
+When a service specify session affinity, ipvs proxy will assign a timeout value(180min by default) to ipvs service.
+
+### Sync period
+
+Similar to iptables proxy, IPVS proxy will do full sync loop every 10 seconds by default. Besides, every update on kubernetes service and endpoint will trigger an ipvs service and destination update.
+
+### Support Only NodeLocal Endpoints
+
+Similar to iptables proxy, IPVS proxy support Only NodeLocal Endpoints feature. When a service has the annotation,  ipvs proxy will do traffic proxy only for local endpoints.
+
+### Network policy
+
+For IPVS NAT mode to work, **all packets from the realservers to the client must go through the director**. It means ipvs proxy should do SNAT in a L3-overlay network(such as flannel) for cross-host network communication. When a container request a cluster IP to visit an endpoint in another host, we should enable `--masquer-all` for ipvs proxy, **which will break network policy**.
```

Comment:
I re-tested in my environment(Flannel VXLAN + IPVS service proxy) and found something different.

There is an ipvs service with 2 destinations which are in the different host.
```
[root@SHA1000130405 home]# ipvsadm -ln
IP Virtual Server version 1.2.1 (size=4096)
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn     
TCP  10.102.128.4:3080 rr
  -> 10.244.0.235:8080            Masq    1      0          0         
  -> 10.244.2.123:8080            Masq    1      0          0 
```

There was no SNAT rules applied and I `curl VIP` in each container.

```
# In container 10.244.0.235

$curl 10.102.128.4:3080
# get response

tcpdumped in another host to see what is source IP is.

$ tcpdump -i flannel.1

20:44:48.021765 IP 10.244.0.235.36844 > 10.244.2.123.webcache: Flags [S], seq 519767460, win 28200, options [mss 1410,sackOK,TS val 416
20:44:48.021998 IP 10.244.2.123.webcache > 10.244.0.235.36844: Flags [S.], seq 1131844123, ack 519767461, win 27960, options [mss 1410,76,nop,wscale 7], length 0
```

So, it seems that **there is no need to do SNAT for cross-host communication**.

However, when ipvs scheduled the self container, there is no response returned. It seems that the packet is dropped when the traffic reached the self container. 

I can confirm that it has nothing to do with `same host vs. cross host`, the issue is about **whether the request hit the same container.**

When I applied SNAT rules as iptables proxy does, it can reach self container via VIP.

```
iptables -t nat -A PREROUTING -s 10.244.0.235 -j KUBE-MARK-MASQ
```

Unfortunately, the source container lost the source container IP and the source IP became flannel.1 IP.


Anyway, it's a good news to me that SNAT is unnecessary for cross-host communication though I need more knowledge to fix the issue I mentioned above.